app/src/main/java/com/example/playlistmaker/presentation/audioplayer/AudioPlayerViewModel.kt

```diff
@@ -1,97 +1,89 @@
-        handler.removeCallbacksAndMessages(SEARCH_REQUEST_TOKEN)
+        StateAudioPlayer.Default()
     }
 
-    fun preparePlayer(url: String) {
-        audioPlayerInteractor.preparePlayer(url = url) { state ->
-            when (state) {
-                StateAudioPlayer.PREPARED -> playEnableState.postValue(true)
-                else -> {}
-            }
+    fun playbackControl() {
+        if (audioPlayerInteractor.getCurrentState() == State.PLAYING) {
+            pausePlayer()
+        } else {
+            startPlayer()
         }
     }
 
-    fun playbackControl(count: Long) {
-        if (count > 0) {
-            audioPlayerInteractor.changingPlayer { state ->
-                when (state) {
-                    StateAudioPlayer.PREPARED, StateAudioPlayer.PAUSED -> {
-                        playState.postValue(true)
-                        handler.removeCallbacks(playerRunnable)
-                    }
-                    StateAudioPlayer.PLAYING -> {
-                        playState.postValue(false)
-                        startTimer(count)
-                    }
-                    else -> {}
-                }
-            }
+    fun preparePlayer(url: String) {
+        audioPlayerInteractor.preparePlayer(url = url) {
+            playState.postValue(StateAudioPlayer.Prepared())
+            timerJob?.cancel()
         }
+        playState.postValue(StateAudioPlayer.Prepared())
     }
 
+
     fun pausePlayer() {
         audioPlayerInteractor.pausePlayer()
-        playState.postValue(false)
-        handler.removeCallbacksAndMessages(SEARCH_REQUEST_TOKEN)
+        playState.postValue(StateAudioPlayer.Paused(getCurrentPlayerPosition()))
+        timerJob?.cancel()
+    }
+
+    private fun startPlayer() {
+        audioPlayerInteractor.startPlayer()
+        playState.postValue(StateAudioPlayer.Playing(getCurrentPlayerPosition()))
+        startTimer()
     }
 
-    fun startTimer(duration: Long) {
-        val startTime = System.currentTimeMillis()
-        playerRunnable =
-            createUpdateTimerTask(startTime, duration * DELAY_MILLIS)
-        handler.post(playerRunnable)
+    private fun getCurrentPlayerPosition(): String {
+        return SimpleDateFormat(
+            "mm:ss",
+            Locale.getDefault()
+        ).format(audioPlayerInteractor.getCurrentPosition()) ?: "00:00"
     }
 
-    fun createUpdateTimerTask(startTime: Long, duration: Long): Runnable {
-        return object : Runnable {
-            override fun run() {
-                val elapsedTime = System.currentTimeMillis() - startTime
-                val remainingTime = duration - elapsedTime
-
-                if (remainingTime > 0) {
-                    secondState.postValue(remainingTime / DELAY_MILLIS)
-                    handler.postDelayed(this, DELAY_MILLIS)
-                } else {
-                    pausePlayer()
-                }
+    private fun startTimer() {
+        timerJob = viewModelScope.launch {
```

Comment:
Лучше отменять старую джобу перед запуском новой корутины на всякий случай:
```suggestion
        timerJob?.cancel()
        timerJob = viewModelScope.launch {
```